### PR TITLE
[chip_sw, kmac_entropy] Checking timeout of EDN to KMAC

### PR DESCRIPTION
**Pull request by ballifatih**
Note: this is the first incremental towards kmac_entropy test, however
there is an issue with the verilator.

This commit includes a simple test where ENTROPY_PERIOD reg is
configured such that:

`prescaler` = 0x3FF
`wait_timer` = 0xFFFF

so that we can observe the maximum timeout value for the EDN request.
The issue is that even though we pick the largest possible values, EDN
still times out (which should not happen).

In the case of FPGA, the EDN request does not time out.

Signed-off-by: Fatih Balli <fatihballi@opentitan.org>

### CODE REVIEW

**Comment by eunchan** [comment]
CC: @weicaiyang @sriyerg 

Should we check if chip level tests are able to run on verilator?

**Review by mwbranstad — COMMENTED** [review]
LGTM, will let others approve.

**Comment by eunchan** [comment]
https://github.com/ballifatih/opentitan/pull/1

Just added a PR to your repo to add the test to dvsim test.

**Comment by ballifatih** [comment]
The current issue is that I am not able to successfully clear the `ERR_CODE` register. @eunchan

File: sw/device/tests/kmac_entropy_test.c
Comment (weicaiyang) [inline sw/device/tests/kmac_entropy_test.c:152]: nit: replace all `%08x` -> `0x%08x`

File: sw/device/tests/kmac_entropy_test.c
Comment (weicaiyang) [inline sw/device/tests/kmac_entropy_test.c:211]: nit: why not also declare the `i` here, so that it's a one-liner
```suggestion
  for (size_t i = 0; i < nr_of_tests; i++) {
```

**Review by weicaiyang — COMMENTED** [review]
please merge all commits to one, don't make a new commit for each update

File: sw/device/tests/kmac_entropy_test.c
Comment (weicaiyang) [inline sw/device/tests/kmac_entropy_test.c:241]: nit: In DV style, we prefer not to use extra delineators like `======`. We don't see we use this kind of comment in C code either.

**Comment by weicaiyang** [comment]
> CC: @weicaiyang @sriyerg
> 
> Should we check if chip level tests are able to run on verilator?

I don't run verilator, but I think chip level tests should be able to run on verilator unless it's in sim_dv which requires a SV component

File: sw/device/tests/kmac_entropy_test.c
Comment (ballifatih) [reply sw/device/tests/kmac_entropy_test.c]: Thanks, I fixed it.

File: sw/device/tests/kmac_entropy_test.c
Comment (ballifatih) [reply sw/device/tests/kmac_entropy_test.c]: Thanks, I fixed it.

File: sw/device/tests/kmac_entropy_test.c
Comment (ballifatih) [reply sw/device/tests/kmac_entropy_test.c]: I was using this to increase visibility in the printed messages, but I removed it for consistency.

**Comment by ballifatih** [comment]
Thanks for the suggestions @weicaiyang :). 

I think we can still keep this PR as a draft until the issue #14713 is fixed. Once, this issue is fixed, I will come back and update the commit again.

**Comment by ballifatih** [comment]
This is still not ready for merge (requires some clean-up + improvement), but I need some suggestions on how to improve kmac_entropy_test. It tested it on dvsim and it works. @eunchan @cindychip @weicaiyang 

Currently, I test whether EDN requests by KMAC times out for six different hard-coded `(prescaler, wait_timer)` pairs. For any pair, the function `is_timeout_expected` compares this pair with a constant `expected_EDN_arrival_time` which acts as a threshold. I set `expected_EDN_arrival_time = 3` by experimenting with dvsim. I am guessing there should be a better way of deciding whether a `(prescaler, wait_timer)` will cause an EDN timeout or not. In other words, do we know how many clock cycles are required for EDN response on average. Any suggestions? 

I also see in the test plan that we should check whether KMAC's hash counter also triggers and EDN request. I am not sure how we can check this on SW.

**Comment by cindychip** [comment]
> This is still not ready for merge (requires some clean-up + improvement), but I need some suggestions on how to improve kmac_entropy_test. It tested it on dvsim and it works. @eunchan @cindychip @weicaiyang
> 
> Currently, I test whether EDN requests by KMAC times out for six different hard-coded `(prescaler, wait_timer)` pairs. For any pair, the function `is_timeout_expected` compares this pair with a constant `expected_EDN_arrival_time` which acts as a threshold. I set `expected_EDN_arrival_time = 3` by experimenting with dvsim. I am guessing there should be a better way of deciding whether a `(prescaler, wait_timer)` will cause an EDN timeout or not. In other words, do we know how many clock cycles are required for EDN response on average. Any suggestions?
> 
> I also see in the test plan that we should check whether KMAC's hash counter also triggers and EDN request. I am not sure how we can check this on SW.

Hi Fatih!

Thanks for working on this test! Here are my thoughts:
1). For entropy timeout threshold, if you want to know the exact cycles for cross clock domain EDN handshake, maybe @tjaychen or @eunchan will know it better. 
For me, a safe value to always timeout is to set pre-scaler to 0, and wait_timer to 1.
Always not timeout, we can set it to a very large number, like all ones, for wait_timer, and pre-scaler. But maybe designer can give a better suggested value to use in the system.

2). For hash counter triggers EDN request. I think it is hard for SW to check EDN request interface. Maybe you can set the threshold to a small value, and check once the counter reaches the threshold, it reset to 0? @eunchan do you think this is sufficient?

Also I would like to confirm the TODO here in the chip testplan: https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/data/chip_testplan.hjson#L2000
@eunchan has this been finalized or we are still waiting for the security feedback?

Thanks,
Cindy

**Comment by eunchan** [comment]
> On Sep 29, 2022, at 9:15 AM, Cindy Chen ***@***.***> wrote:
> 
> 
> This is still not ready for merge (requires some clean-up + improvement), but I need some suggestions on how to improve kmac_entropy_test. It tested it on dvsim and it works. @eunchan <https://github.com/eunchan> @cindychip <https://github.com/cindychip> @weicaiyang <https://github.com/weicaiyang>
> Currently, I test whether EDN requests by KMAC times out for six different hard-coded (prescaler, wait_timer) pairs. For any pair, the function is_timeout_expected compares this pair with a constant expected_EDN_arrival_time which acts as a threshold. I set expected_EDN_arrival_time = 3 by experimenting with dvsim. I am guessing there should be a better way of deciding whether a (prescaler, wait_timer) will cause an EDN timeout or not. In other words, do we know how many clock cycles are required for EDN response on average. Any suggestions?
> 
> I also see in the test plan that we should check whether KMAC's hash counter also triggers and EDN request. I am not sure how we can check this on SW.
> 
> Hi Fatih!
> 
> Thanks for working on this test! Here are my thoughts:
> 1). For entropy timeout threshold, if you want to know the exact cycles for cross clock domain EDN handshake, maybe @tjaychen <https://github.com/tjaychen> or @eunchan <https://github.com/eunchan> will know it better.
> For me, a safe value to always timeout is to set pre-scaler to 0, and wait_timer to 1.
> Always not timeout, we can set it to a very large number, like all ones, for wait_timer, and pre-scaler. But maybe designer can give a better suggested value to use in the system.
> 
EDN latency differs based on the test environment. Verilator, esp., returns the entropy pretty quickly as far as I know. In real silicon, AFAIK, EDN max latency could be up to 5ms. So, I think @cindychip suggestion looks good.
> 2). For hash counter triggers EDN request. I think it is hard for SW to check EDN request interface. Maybe you can set the threshold to a small value, and check once the counter reaches the threshold, it reset to 0? @eunchan <https://github.com/eunchan> do you think this is sufficient?
> 
I think it is good. Thanks for the suggestion.

**Comment by ballifatih** [comment]
Thanks for the comments @cindychip and @eunchan. I updated the test accordingly and it checks both EDN timeouts and hash counter rewinds after exceeding the threshold. Passes both in verilator and dvsim. Please take a look.

There are also few updates on the `dif_kmac.c`, and it would be nice to get a review on those.

File: sw/device/tests/kmac_entropy_test.c
Comment (cindychip) [inline sw/device/tests/kmac_entropy_test.c:198]: If there is no timeout error, can we add some checks (if not already):
to make sure there is no error or interrupts?
Caz I believe line 122 only checks if the err_code matches the timeout error code.
What if (very unlikely) there are some other errors?

**Review by cindychip — COMMENTED** [review]
The C code logic looks great, just added one minor suggestions.
I am not an expert on C code. I think this is good to go if other SW engineers have already taken a look :)

File: sw/device/lib/dif/dif_kmac.h
Comment (matutem) [inline sw/device/lib/dif/dif_kmac.h:678]: For the stale value detection to work this assumes the INTR_STATE registerr will be cleared after any interrupt, and that interrupts were enabled.

**Review by matutem — COMMENTED** [review]
I left some questions.

**Comment by cindychip** [comment]
@ballifatih I think the CI fpga test failed on: //sw/device/lib/dif:kmac_unittest                               FAILED TO BUILD
Please feel free to let me know if you have any questions :)

**Comment by ballifatih** [comment]
> @ballifatih I think the CI fpga test failed on: //sw/device/lib/dif:kmac_unittest FAILED TO BUILD Please feel free to let me know if you have any questions :)

Thanks @cindychip. Yes, there was an issue with `dif_kmac_config_is_locked`, which is fixed in PR #15431. I just rebased, and this should solve the CI failure.

**Comment by ballifatih** [comment]
Currently the test fails because the EDN request doesn't time out on FPGA. In particular, even for the smallest pair (`prescaler = 0x0`, `wait_timer = 0x001`), a timeout error is not generated for FPGA.

The test works both on verilator and dvsim (timeout happens as expected).

File: sw/device/tests/BUILD
Comment (moidx) [inline sw/device/tests/BUILD:880]: Add the following to skip the FPGA target for now:
```bazel
    cw310 = cw310_params(
        # FIXME #15530: EDN doesn't timeout on FPGA 
        tags = ["broken"],
    ),
```

**Review by cindychip — APPROVED** [review]
LGTM, thanks for working on this test!

File: sw/device/tests/kmac_entropy_test.c
Comment (moidx) [inline sw/device/tests/kmac_entropy_test.c:133]: The cast is not required here.

File: sw/device/tests/kmac_entropy_test.c
Comment (moidx) [inline sw/device/tests/kmac_entropy_test.c:178]: nit: `/*processed=*/NULL` to follow the style guide.

**Review by moidx — APPROVED** [review]
Thanks @ballifatih! Approving with some C style guide nits. Once you have updated the commit, plmk so that we can submit this change.

**Comment by ballifatih** [comment]
Thanks for the nit comments @moidx , I just made these changes. The only thing that is left is `static_assert` for the hard-coded value of `kDifErrorEntropyWaitTimerExpired`, because nothing equivalent is exposed in `kmac_regs.h`.

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:56]: Yes, this type doesn't need a name, good job!

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:80]:
```suggestion
static const kmac_test_t kKmacTestVector = {
```

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:141]:
```suggestion
  dif_kmac_customization_string_t *str_buffer_ptr =
```
or something along these lines.

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:144]:
```suggestion
  size_t digest_len =
```
for consistency

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:147]:
```suggestion
  for (size_t i = 0; i < kKmacTimeoutTestCount; ++i) {
```

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:151]:
```suggestion
    kmac_edn_timeout_t timeout_test  = kTestTimeoutVals[i];
```

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:174]:
```suggestion
    dif_result_t res =
        dif_kmac_squeeze(&kmac, &kmac_operation_state, out,
                         kmac_test_vector.digest_len, /*processed=*/ NULL);
```

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:186]: Two questions: 1)  Can `err_status` be `kDifErrorNone`? 2) Can there be more than one error pending? If so, should this be a loop?

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [inline sw/device/tests/kmac_entropy_test.c:197]: nit: Can we move the contents of the `if` block into the block above and add the `else` block after that?

**Review by alphan — APPROVED** [review]
Thanks @ballifatih! Please see my comments (mostly nits).

**Comment by ballifatih** [comment]
Thanks for the detailed review @alphan.

File: sw/device/lib/dif/dif_kmac.h
Comment (cindychip) [reply sw/device/lib/dif/dif_kmac.h:383]: @ballifatih We talked offline and yeah it is not possible to add the assertion because it is not a enum on the register table. Could you add a comment saying it is coming from the RTL package?
But this is not blocking, we can enhance it in next PR ;)

**Comment by alphan** [comment]
> Thanks for the detailed review @alphan. I made most of these changes, except the ones I dropped comments above.

Thanks @ballifatih but I don't see any replies from you. As long as the error checking behavior is correct the rest of the comments were small nits.

File: sw/device/lib/dif/dif_kmac.h
Comment (ballifatih) [reply sw/device/lib/dif/dif_kmac.h]: When I was updating the test, I had written some responses to your questions, as I updated the test. However, it turns out that I had left them on "pending" state so they were actually never visible to you. Sorry about this, and thanks for the review @matutem

**Comment by ballifatih** [comment]
> 
> Thanks @ballifatih but I don't see any replies from you. As long as the error checking behavior is correct the rest of the comments were small nits.

Thanks @alphan, you just made me realize that I was actually leaving my responses in "pending" status, instead of properly finishing the reviews! I hope the responses are visible now.

**Comment by alphan** [comment]
> > Thanks @ballifatih but I don't see any replies from you. As long as the error checking behavior is correct the rest of the comments were small nits.
> 
> Thanks @alphan, you just made me realize that I was actually leaving my responses in "pending" status, instead of properly finishing the reviews! I hope the responses are visible now.

That's what I suspected, I can see them now, thanks!

File: sw/device/tests/kmac_entropy_test.c
Comment (alphan) [reply sw/device/tests/kmac_entropy_test.c]: Thanks! We can update other files when we touch them, no need to change everything at once. We should also codify this in the style guide if it's not already there.

**Comment by alphan** [comment]
Your latest changes LGTM, resolved my remaining comments. Thanks!

**Pull request merged**
